drop-storage/src/lib.rs

```diff
@@ -32,4 +35,671 @@ impl Storage {
+            started_states: vec![],
+            cancel_states: vec![],
+            failed_states: vec![],
+            completed_states: vec![],
+        })
+        .collect::<Vec<_>>();
+
+        for path in &mut paths {
+            path.pending_states = sqlx::query!(
+                "SELECT * FROM outgoing_path_pending_states WHERE path_id = ?1",
+                path.id
+            )
+            .fetch_all(&mut conn)
+            .await
+            .map_err(error::Error::DBError)?
+            .iter()
+            .map(|s| OutgoingPathPendingState {
+                path_id: s.path_id,
+                created_at: s.created_at.timestamp_millis(),
+            })
+            .collect();
+
+            path.started_states = sqlx::query!(
+                "SELECT * FROM outgoing_path_started_states WHERE path_id = ?1",
+                path.id
+            )
+            .fetch_all(&mut conn)
+            .await
+            .map_err(error::Error::DBError)?
+            .iter()
+            .map(|s| OutgoingPathStartedState {
+                path_id: s.path_id,
+                bytes_sent: s.bytes_sent,
+                created_at: s.created_at.timestamp_millis(),
+            })
+            .collect();
+
+            path.cancel_states = sqlx::query!(
+                "SELECT * FROM outgoing_path_cancel_states WHERE path_id = ?1",
+                path.id
+            )
+            .fetch_all(&mut conn)
+            .await
+            .map_err(error::Error::DBError)?
+            .iter()
+            .map(|s| OutgoingPathCancelState {
+                path_id: s.path_id,
+                by_peer: s.by_peer,
+                bytes_sent: s.bytes_sent,
+                created_at: s.created_at.timestamp_millis(),
+            })
+            .collect();
+
+            path.failed_states = sqlx::query!(
+                "SELECT * FROM outgoing_path_failed_states WHERE path_id = ?1",
+                path.id
+            )
+            .fetch_all(&mut conn)
+            .await
+            .map_err(error::Error::DBError)?
+            .iter()
+            .map(|s| OutgoingPathFailedState {
+                path_id: s.path_id,
+                status_code: s.status_code,
+                bytes_sent: s.bytes_sent,
+                created_at: s.created_at.timestamp_millis(),
+            })
+            .collect();
+
+            path.completed_states = sqlx::query!(
+                "SELECT * FROM outgoing_path_completed_states WHERE path_id = ?1",
+                path.id
+            )
+            .fetch_all(&mut conn)
+            .await
+            .map_err(error::Error::DBError)?
+            .iter()
+            .map(|s| OutgoingPathCompletedState {
+                path_id: s.path_id,
+                created_at: s.created_at.timestamp_millis(),
+            })
+            .collect();
+        }
+
+        Ok(paths)
+    }
+
+    async fn get_incoming_paths(&self, transfer_id: String) -> Result<Vec<IncomingPath>> {
```

Comment:
This whole file really begs for unit tests. it would be great to simulate insertions of a few transfers with few files, imitate several fails, cancels and etc(really all the calls) and except the proper data to be returned.

Reply:
Created a ticket for this